### PR TITLE
[Solver] XMPM

### PR DESCRIPTION
**Pull request by yliang-sn**
**Describe the PR**
A clear and concise description of the PR.

**Related Issues/PRs**

**Additional context**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -146,6 +146,7 @@ include_directories(BEFORE
   ${mpm_SOURCE_DIR}/include/solvers/mpm_scheme/
   ${mpm_SOURCE_DIR}/external/
   ${mpm_SOURCE_DIR}/tests/include/
+  ${mpm_SOURCE_DIR}/include/xmpm/
```
Comment (bodhinandach) [inline CMakeLists.txt:149]: test

**Review by bodhinandach — CHANGES_REQUESTED** [review]
Please change this

File: include/mesh.h [hunk 2]
```diff
@@ -500,6 +501,78 @@ class Mesh {
   // Initialise the nodal properties' map
   void initialise_nodal_properties();
 
+  //! Create the nodal properties' map for discontinuity
```
Comment (bodhinandach) [inline include/mesh.h:504]: Please group these functions

File: include/node.h [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #include "mutex.h"
 #include "nodal_properties.h"
 #include "node_base.h"
+#include <fstream>
```
Comment (bodhinandach) [inline include/node.h:8]: Please remove this

File: include/node.h [hunk 4]
```diff
@@ -47,6 +48,13 @@ class Node : public NodeBase<Tdim> {
       unsigned prop_id,
       std::shared_ptr<mpm::NodalProperties> property_handle) noexcept override;
 
+  //! Initialise shared pointer to nodal properties pool for discontinuity
```
Comment (bodhinandach) [inline include/node.h:51]: Please group these functions

File: include/node.h [hunk 5]
```diff
@@ -577,6 +656,11 @@ class Node : public NodeBase<Tdim> {
   std::unique_ptr<spdlog::logger> console_;
   //! MPI ranks
   std::set<unsigned> mpi_ranks_;
+  //! discontinuity enrich
```
Comment (bodhinandach) [inline include/node.h:659]: Please group the variables into xmpm variables

File: include/node_base.h [hunk 6]
```diff
@@ -217,12 +224,15 @@ class NodeBase {
   //! Assign velocity constraint
   //! Directions can take values between 0 and Dim * Nphases
   //! \param[in] dir Direction of velocity constraint
-  //! \param[in] velocity Applied velocity constraint
+  //! \param[ivirtualn] velocity Applied velocity constraint
```
Comment (bodhinandach) [inline include/node_base.h:227]: remove typo

File: include/node.tcc [hunk 7]
```diff
@@ -38,6 +38,8 @@ void mpm::Node<Tdim, Tdof, Tnphases>::initialise() noexcept {
   status_ = false;
   solving_status_ = false;
   material_ids_.clear();
+  // mark: need to check
+  discontinuity_enrich_ = false;
```
Comment (bodhinandach) [inline include/node.tcc:42]: Make initialise_xmpm or whatever. See `initialise_twophase()`

File: include/particles/particle_base.h [hunk 8]
```diff
@@ -347,6 +347,41 @@ class ParticleBase {
   virtual void deserialize(
       const std::vector<uint8_t>& buffer,
       std::vector<std::shared_ptr<mpm::Material<Tdim>>>& materials) = 0;
+  //! set the level set function values
```
Comment (bodhinandach) [inline include/particles/particle_base.h:350]: Make error throw in base functions

File: include/particles/pod_particles/pod_particle.h [hunk 9]
```diff
@@ -44,6 +44,8 @@ typedef struct PODParticle {
   unsigned material_id;
   // Number of state variables
   unsigned nstate_vars;
+  //! Level set values
```
Comment (bodhinandach) [inline include/particles/pod_particles/pod_particle.h:47]: Let's derive a new pod class for this

File: include/solvers/mpm_base.tcc [hunk 10]
```diff
@@ -27,6 +27,12 @@ mpm::MPMBase<Tdim>::MPMBase(const std::shared_ptr<IO>& io) : mpm::MPM(io) {
       {"mass", VariableType::Scalar},
       {"volume", VariableType::Scalar},
       {"mass_density", VariableType::Scalar},
+      {"levelset", VariableType::Scalar},
+      {"minimum_acoustic_eigenvalue", VariableType::Scalar},
+      {"discontinuity_angle", VariableType::Scalar},
+      {"first_principal_stress", VariableType::Scalar},
+      {"first_principal_strain", VariableType::Scalar},
+      {"energy", VariableType::Scalar},
```
Comment (bodhinandach) [inline include/solvers/mpm_base.tcc:35]: Have a look and tidy up

File: include/node.tcc (hunk 7 above)
Comment (yliang-sn) [reply include/node.tcc:42]: will do it in the derived XMPM node class

**Comment by bodhinandach** [comment]
Closing this temporarily until you are ready with the new version of XMPM @yliang-sn

**Pull request closed without merging**